Added Attempt History Table for each User and Account Deletion Functionality

Kidsnd274 opened the pull request:
## Changes Made:
- Implement Account Delete Functionality (both password & Google accounts)
- Implement User Attempt History Table in each User's Profile Page
  - Connect User Service to Fire Store
  - Fire Store now stores attempt history
  - Matching Service updates User Service when matches occur
- Refactor verify_token functions in User Service
- Remove unnecessary refreshUser call from Navbar
- Fixed bug where the login page doesn't load on the first load (because of Navbar trying to access the profile picture while the user is logged out)

## More Info
Firestore now stores attempt history for each user. When a match occurs, matching service tells user service who the matched users are, the time and the question.

User Service stores this match in Firestore in this manner.
attempt_history -> (user_id) -> attempts -> (session_id)
With this format:
```
matched_user: "user_id" (string)
question_id: "question_id" (string)
timestamp: "1729500216.228071": (string)
```
So when a match occurs, User Service creates two new entries for both of the matched users involved.

## How to test:
Login to two instances of PeerPrep, and start a match between both. When a match is found, the attempt history in each of the user's profile page should be updated with the latest match.

Fixes #7

Kidsnd274 committed:
Implement Delete Account (password only)

Kidsnd274 committed:
Fixed bug with profile settings button, and form bugs with Delete Acc Pass component

Kidsnd274 committed:
Initial Implementation of Google Acc Delete

There are issues trying to reauthenticate with Google. For now ignored

Kidsnd274 committed:
Refactor verify_token functions to allow reuse

Kidsnd274 committed:
Connect firestore to user service

Kidsnd274 committed:
Initial Implementation of User Attempt History

Kidsnd274 committed:
Add Bruno scripts for API testing

Kidsnd274 committed:
Initial Implementation of Attempt History Table

Kidsnd274 committed:
Improved Attempt History Table with more information

Kidsnd274 committed:
Convert Epoch Time to Date Time

Kidsnd274 committed:
Capitalize first letter in Difficulty label

yfg-dev committed:
Send matched users info to user service

Kidsnd274 committed:
Fixed bug causing first load to not work

Navbar tries to access profile picture when Uesr is not logged in yet

Kidsnd274 committed:
Sort history by timestamp descending order

Kidsnd274 committed:
Remove unnecessary refreshUser call from Navbar

Kidsnd274 committed:
Fix issue where Attempt History doesn't load if matched user doesn't exist

Account may have been deleted

Kidsnd274 committed:
Fix bug causing matched user display name to not show up

Kidsnd274 committed:
Require token when request for attempt history

Kidsnd274 committed:
Remove bruno scripts

yfg-dev committed:
Saved question as attempted only when timer is still counting down

yfg-dev reviewed (changes requested):
- question attempt histories display correctly for partial and complete match.
- correctly update display name for both google account and email account
- correct name displayed in attempt history when matched user changed display name.
- handle the case when the matched user account is deleted and displayed unknown user for the matched user field.
- able to delete email account.
- handled the case of incorrect password when wrong password is entered.

failed to delete google account
![Screenshot from 2024-10-25 19-02-20](https://github.com/user-attachments/assets/e75a7c15-0ceb-4de3-a0de-a84ccfbb0d13)

Kidsnd274 committed:
Used another method to delete Google Account

Used User Service to execute command instead due to some issue with the reauthenticateWithPopup function

Kidsnd274 committed:
Merge branch 'question-attempt-history' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g29 into question-attempt-history

yfg-dev reviewed (approved):
LGTM
